### PR TITLE
[prototype] migration: onboard the first long-running migration

### PR DESCRIPTION
**Pull request by irfansharif**
This PR onboards the first real long-running migration using the
infrastructure we've been building up within pkg/migration. It adds in
the final missing pieces described in our original RFC (#48843). These
components were originally prototyped in #56107.

The migration in question (which happens to be a below-Raft one, first
attempted in #42822) now lets us do the following:

  i.  Use the RangeAppliedState on all ranges
  ii. Use the unreplicated TruncatedState on all ranges

The missing pieces we introduce along side this migration are:

  a. The `Migrate` KV request. This forces all ranges overlapping with
  the request spans to execute the (below-raft) migrations corresponding
  to the specific version, moving out of any legacy modes they may
  currently be in. KV waits for this command to durably apply on all the
  replicas before returning, guaranteeing to the caller that all
  pre-migration state has been completely purged from the system.

  b. `IterateRangeDescriptors`. This provides a handle on every range
  descriptor in the system, which callers can then use to send out
  arbitrary KV requests to in order to run arbitrary KV-level
  migrations. These requests will typically just be the `Migrate`
  request, with added code next to the `Migrate` command implementation
  to do the specific KV-level things intended for the specified version.

  c. The `system.migrations` table. We use it to store metadata about
  ongoing migrations for external visibility/introspection. The schema
  (listed below) is designed with an eye towards scriptability. We
  want operators to be able programmatically use this table to control
  their upgrade processes, if needed.

      CREATE TABLE public.migrations (
          version STRING NOT NULL,
          status STRING NOT NULL,
          description STRING NOT NULL,
          start TIMESTAMP NOT NULL DEFAULT now():::TIMESTAMP,
          completed TIMESTAMP NULL,
          progress STRING NULL,
          CONSTRAINT "primary" PRIMARY KEY (version ASC),
          FAMILY "primary" (version, status, description, start, completed, progress)
      )

Release note(general change): Cluster version upgrades, as initiated by
`SET CLUSTER SETTING version = <major>-<minor>`, now perform internal
maintenance duties that will delay how long it takes for the command to
complete. The delay is proportional to the amount of data currently
stored in the cluster. The cluster will also experience a small amount
of additional load during this period while the upgrade is being
finalized.

Release note(general change): We introduce a new `system.migrations`
table for introspection into crdb internal data migrations. These
migrations are the "maintenance duties" mentioned above. The table
surfaces the currently ongoing migrations, the previously completed
migrations, and in the case of failure, the errors from the last failed
attempt.

### CODE REVIEW

**Comment by irfansharif** [comment]
Heads up, this is still a [wip] branch. It's a bit far from landing. I have zero tests, and intend to add them here. I also intend to break it up into a few commits: introducing the system table, introducing the request type, introducing IterateRangeDescriptors, and only after all that, adding the migration itself. For now I've just squashed them for an early first-pass look/sanity check before I spend more time on final polish.

I have a few open questions in there as comments (grep for `XXX:`), but any+all thoughts welcome on the rest of it.

**Review by tbg — COMMENTED** [review]
This looks pretty good already! I left a bunch of comments but I think it's basically time to figure out how to write proper tests and how to put this into a number of smaller PRs.

The migration table introduction looks like it could be a good last PR. The migrate command + testing could come first. Then a PR that updates the manager and adds tests and hooks up the migrate command. Just a suggestion.

> Reviewed 49 of 49 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57445#-:-MNceNd_E8feLkVt0rOK:b-l2xnil)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif and @knz)

___
*[pkg/kv/kvserver/replica_proposal.go, line 781 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57445#-MNcS44WCptn6sZ3SmNw:-MNcS44X2vlnrdzvD6lG:b-cdneo2) ([raw file](https://github.com/cockroachdb/cockroach/blob/45ce792773f010ca42422781ab40da54868c1c58/pkg/kv/kvserver/replica_proposal.go#L781)):*
> ```Go
> 			// below. If we're not running 21.1 yet, migrate over as we've done
> 			// since the introduction of the applied state key.
> 			if r.ClusterSettings().Version.IsActive(ctx, clusterversion.TruncatedAndRangeAppliedStateMigration) {
> ```

This is exactly the kind of thing that could cause spurious fatal errors when there are replicaGC'able replicas left around that didn't get included in the raft command. On these, you can still hit the evaluation path (they'll try to request leases). I don't know if this PR already does anything to avoid this behavior (I'll just leave this comment now and will continue reviewing then), but as we've discussed in the past there are two ingredients:

- track the range version (in the replica state) and disallow snapshots from an old version to any newer version
- force a run of replicaGC on all replicas in the cluster that don't have the updated range version (via EveryNode presumably)

That way we can

1. call Migrate on the whole keyspace
2. force the run

and be assured that all replicas now in the cluster are at the new version, i.e. it's safe to bump the cluster version.
___
*[pkg/kv/kvserver/replica_write.go, line 223 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57445#-MNccg5V8aYyKDVIRTho:-MNccg5WCnA-6RbikuRO:b-l8p4zh) ([raw file](https://github.com/cockroachdb/cockroach/blob/45ce792773f010ca42422781ab40da54868c1c58/pkg/kv/kvserver/replica_write.go#L223)):*
> ```Go
> 				// applied on all peers, which we achieve via waitForApplication.
> 				//
> 				// XXX: Could a snapshot below maxLeaseIndex be in-flight to a
> ```

Snapshots that don't include the recipient (as specified by replicaID and descriptor in the snap vs the replicaID of the raft instance) are discarded by the recipient, so this should not be an issue. The particular case you mention will have the learner have a higher replicaID than the incoming snapshot, so the snapshot will also be out for that reason. Finally, a snapshot is discarded unless it moves the LAI forward.
___
*[pkg/kv/kvserver/batcheval/cmd_migrate.go, line 16 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57445#-MNcVQ5K89TCiMJ7ZJ6l:-MNcVQ5K89TCiMJ7ZJ6m:b3hz2kb) ([raw file](https://github.com/cockroachdb/cockroach/blob/45ce792773f010ca42422781ab40da54868c1c58/pkg/kv/kvserver/batcheval/cmd_migrate.go#L16)):*
> ```Go
> 	"context"
> 
> 	cv "github.com/cockroachdb/cockroach/pkg/clusterversion"
> ```

small nit, but I'm generally -1 on unnecessary import renames.
___
*[pkg/migration/helper.go, line 39 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57445#-MNcWErx40DIgQFQYsmd:-MNcWErx40DIgQFQYsme:b62crl6) ([raw file](https://github.com/cockroachdb/cockroach/blob/45ce792773f010ca42422781ab40da54868c1c58/pkg/migration/helper.go#L39)):*
> ```Go
> // to execute all relevant node-level operations on them. If any of the nodes
> // are found to be unavailable, an error is returned.
> func (h *Helper) RequiredNodeIDs(ctx context.Context) ([]roachpb.NodeID, error) {
> ```

Should this be exported given that you need to wrap it in a stabilizing retry loop, and thus is a footgun if used from a migration work function?
___
*[pkg/migration/helper.go, line 100 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57445#-MNcVx_jASeI1T3qIuEK:-MNcVx_jASeI1T3qIuEL:b-79vu4x) ([raw file](https://github.com/cockroachdb/cockroach/blob/45ce792773f010ca42422781ab40da54868c1c58/pkg/migration/helper.go#L100)):*
> ```Go
> // (accessible using the `progress` column in `system.migrations`) for the
> // ongoing migration.
> func (h *Helper) UpdateProgress(ctx context.Context, progress string) error {
> ```

"progress" is a bit of a misnomer, how about "status"
___
*[pkg/migration/helper.go, line 136 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57445#-MNcWQvTDy965IA0ZVB_:-MNcWQvTDy965IA0ZVBa:b4ehhab) ([raw file](https://github.com/cockroachdb/cockroach/blob/45ce792773f010ca42422781ab40da54868c1c58/pkg/migration/helper.go#L136)):*
> ```Go
> //
> // By the time EveryNode returns, we'll have thus invoked the closure against
> // every node in the cluster.
> ```

That's not true. A node can always join just as EveryNode returns. What we have achieved is that there is now a causal happened-before relation between executing something on every node in the cluster and the next node joining the cluster. We managed to run something on all nodes without a new node joining half-way through, which could have allowed it to pick up some state (off one of the existing nodes that hadn't finished processing yet).

For the example of bumping the cluster version, we'll have achieved proving that by the time EveryNode returns, every node in the cluster, current and future, will use the new cluster version.
___
*[pkg/migration/helper.go, line 162 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57445#-MNcY9e59hjUsfZG6S5L:-MNcY9e59hjUsfZG6S5M:b-5epu3z) ([raw file](https://github.com/cockroachdb/cockroach/blob/45ce792773f010ca42422781ab40da54868c1c58/pkg/migration/helper.go#L162)):*
> ```Go
> 
> 	for {
> 		// TODO(irfansharif): We can/should send out these RPCs in parallel.
> ```

We should do that soon, too. Ideally in the burst of commits coming out of this one!
___
*[pkg/migration/migrations.go, line 110 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57445#-MNcZyAc6dUxb2W-AYW1:-MNcZyAc6dUxb2W-AYW2:bukktt7) ([raw file](https://github.com/cockroachdb/cockroach/blob/45ce792773f010ca42422781ab40da54868c1c58/pkg/migration/migrations.go#L110)):*
> ```Go
> func TruncatedStateMigration(ctx context.Context, h *Helper) error {
> 	// TODO(irfansharif): What should the default block size be?
> 	const blockSize = 50
> ```

At 50k ranges, with each range taking ~200ms (a rather conservative assumption but in a wide-area cluster could be true due to hops from the manager to the replicas) to process the Migrate, a block size of 1 is >2.7 hours. Block size 50 is ~3.3 minutes. Block size 200 should easily be in the cards given that each range is doing very little work here, so down to 50s in this case which seems reasonable.
___
*[pkg/migration/migrations.go, line 116 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57445#-MNcdZR6AF0RidVkLMje:-MNcdZR74QRbeM4tsDpF:bpd2kam) ([raw file](https://github.com/cockroachdb/cockroach/blob/45ce792773f010ca42422781ab40da54868c1c58/pkg/migration/migrations.go#L116)):*
> ```Go
> 		for _, desc := range descriptors {
> 			if bytes.Compare(desc.StartKey, keys.LocalMax) < 0 {
> 				// XXX: Does it make sense to "migrate" ranges with local keys?
> ```

Yeah, this is a bit of a wart. We do want to reach the first range, but we can't address StartKey. However, `LocalMax` is on r1, so in this case we basically just use  `LocalMax` instead of `StartKey`.

I wonder if we'd get sanity back if we officially let `r1` begin at `LocalMax`, but that is not worth getting into here or probably ever.
___
*[pkg/migration/migrations.go, line 128 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57445#-MNc_lsa-n2-rAiTDd43:-MNc_lsa-n2-rAiTDd44:b-a3j89j) ([raw file](https://github.com/cockroachdb/cockroach/blob/45ce792773f010ca42422781ab40da54868c1c58/pkg/migration/migrations.go#L128)):*
> ```Go
> 
> 		numMigratedRanges += len(descriptors)
> 		progress := fmt.Sprintf("[batch %d/??] migrated %d ranges", batchIdx, numMigratedRanges)
> ```

Doesn't `IterateRangeDescriptors` have to handle txn retries? The closure here would need to be idempotent, but the counters aren't. Maybe `IterateRangeDescriptors` should take an `init func()` as well which can reset them.
___
*[pkg/migration/migrations.go, line 140 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57445#-MNcaCI_E3GtY7kubL0W:-MNcaCI_E3GtY7kubL0X:b-70n4re) ([raw file](https://github.com/cockroachdb/cockroach/blob/45ce792773f010ca42422781ab40da54868c1c58/pkg/migration/migrations.go#L140)):*
> ```Go
> 
> 	// XXX: I think we'll need to make sure all stores have synced once to
> 	// persist all applications of the `Migrate` raft command.
> ```

Very good point. It sounds reasonable to do this as a separate round of EveryNode (as opposed to doing it on each raft command application, which would cause tons of syncs).
It's probably also just generally a good idea to make this happen on every migration that does work. It could even be part of bumping the cluster version, which by definition already requires syncing; we could just extend that to explicitly mandate syncing all engines.

Hmm, thinking about this further, this is not good enough when nodes restart rapidly. I think the best way to address this is to make `EveryNode` more aggressive about retrying: in addition to the list of nodes stabilizing, their corresponding epochs also need to remain stable (and, of course, everyone live). That way we don't impose a requirement that everything we do needs to be synced immediately, which is super expensive and we're also pretty much guaranteed to get it wrong.
___
*[pkg/roachpb/api.go, line 1295 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57445#-MNcb-yv634wrZSR5Yqc:-MNcb-yv634wrZSR5Yqd:b9lv34g) ([raw file](https://github.com/cockroachdb/cockroach/blob/45ce792773f010ca42422781ab40da54868c1c58/pkg/roachpb/api.go#L1295)):*
> ```Go
> func (*AdminScatterRequest) flags() int                  { return isAdmin | isRange | isAlone }
> func (*AdminVerifyProtectedTimestampRequest) flags() int { return isAdmin | isRange | isAlone }
> func (*MigrateRequest) flags() int                       { return isWrite | isRange | isAlone }
> ```

I wonder if `MigrateRange` is a better name.
___
*[pkg/roachpb/api.proto, line 1500 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57445#-MNcb7X63Qjck40wzr03:-MNcb7X63Qjck40wzr04:ba5k2ed) ([raw file](https://github.com/cockroachdb/cockroach/blob/45ce792773f010ca42422781ab40da54868c1c58/pkg/roachpb/api.proto#L1500)):*
> ```protobuf
> // is already going to be the one responsible for stepping through all
> // intermediate versions, it's probably saner to only carry out the specific
> // migration associated with the target version.
> ```

Yeah, that's the impression I had as well.
___
*[pkg/roachpb/api.proto, line 1502 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57445#-MNcbBaj0hGZ0lVNSZJH:-MNcbBaj0hGZ0lVNSZJI:b-tdxb80) ([raw file](https://github.com/cockroachdb/cockroach/blob/45ce792773f010ca42422781ab40da54868c1c58/pkg/roachpb/api.proto#L1502)):*
> ```protobuf
> // migration associated with the target version.
> //
> // TODO(irfansharif): Should this be an admin request instead? There's no good
> ```

No, admin requests are really just a way to invoke random code on the leaseholder. Here we want to properly go through the raft pipeline, which admin requests don't.
___
*[pkg/roachpb/api.proto, line 1509 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57445#-MNcbQ8A8DFqcvncB22-:-MNcbQ8A8DFqcvncB220:bus13j7) ([raw file](https://github.com/cockroachdb/cockroach/blob/45ce792773f010ca42422781ab40da54868c1c58/pkg/roachpb/api.proto#L1509)):*
> ```protobuf
>   RequestHeader header = 1 [(gogoproto.nullable) = false, (gogoproto.embed) = true];
> 
>   // The version that will become active next. A request with the
> ```

incomplete
___
*[pkg/roachpb/method.go, line 161 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57445#-MNcbWBsEGLeAVKIJ5yI:-MNcbWBsEGLeAVKIJ5yJ:b-npf4lg) ([raw file](https://github.com/cockroachdb/cockroach/blob/45ce792773f010ca42422781ab40da54868c1c58/pkg/roachpb/method.go#L161)):*
This comment could be better, something like

> Migrate updates the range state to conform to a specified cluster version. It is our main mechanism for phasing out legacy code below Raft.
___
*[pkg/sql/catalog/systemschema/system.go, line 347 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57445#-MNcbpUc-mvB_HbSAbsX:-MNcbpUc-mvB_HbSAbsY:b80yf79) ([raw file](https://github.com/cockroachdb/cockroach/blob/45ce792773f010ca42422781ab40da54868c1c58/pkg/sql/catalog/systemschema/system.go#L347)):*
> ```Go
> 
> 	MigrationsTableSchema = `
> CREATE TABLE system.migrations (
> ```

Is this going to be gossiped? (The answer should be "no", I forget what the rules are).
___
*[pkg/sql/catalog/systemschema/system.go, line 615 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57445#-MNcbyq26CWNL3LHNM_-:-MNcbyq26CWNL3LHNM_0:b-h63sr2) ([raw file](https://github.com/cockroachdb/cockroach/blob/45ce792773f010ca42422781ab40da54868c1c58/pkg/sql/catalog/systemschema/system.go#L615)):*
> ```Go
> 
> 	// MigrationsTable is the descriptor for the migrations table. It contains
> 	// details for all past and ongoing migrations.
> ```

We'll want someone from SQL to review this part of the PR. This all looks sane, but I'd want someone to sign off that it is.
___
*[pkg/sql/logictest/testdata/logic_test/distsql_agg, line 578 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57445#-MNcc385945SOpGtv5Dw:-MNcc385945SOpGtv5Dx:bmyo2pb) ([raw file](https://github.com/cockroachdb/cockroach/blob/45ce792773f010ca42422781ab40da54868c1c58/pkg/sql/logictest/testdata/logic_test/distsql_agg#L578)):*
> ```
> SELECT corr(y, x)::decimal, covar_pop(y, x)::decimal, covar_samp(y, x)::decimal FROM statistics_agg_test
> ----
> 0.04522896319136315  3.75  3.787878787878788
> ```

Huh. What are these about?

**Review by irfansharif — COMMENTED** [review]
Thanks for looking. I've addressed all open comments here, and added a bunch of tests. I'm still going to leave this up as an open prototype branch (for this top level review + discussion + running CI) and pull out bits into separate PRs elsewhere.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57445#-:-MNxc6ouC7m4XqGg9S7Z:btrw53g)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz, @miretskiy, @tbg, and @yuzefovich)

___
*[pkg/kv/kvserver/replica_proposal.go, line 781 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57445#-MNcS44WCptn6sZ3SmNw:-MNqF86Y9OwwhUa-IO3E:brls3ya) ([raw file](https://github.com/cockroachdb/cockroach/blob/45ce792773f010ca42422781ab40da54868c1c58/pkg/kv/kvserver/replica_proposal.go#L781)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

This is exactly the kind of thing that could cause spurious fatal errors when there are replicaGC'able replicas left around that didn't get included in the raft command. On these, you can still hit the evaluation path (they'll try to request leases). I don't know if this PR already does anything to avoid this behavior (I'll just leave this comment now and will continue reviewing then), but as we've discussed in the past there are two ingredients:

- track the range version (in the replica state) and disallow snapshots from an old version to any newer version
- force a run of replicaGC on all replicas in the cluster that don't have the updated range version (via EveryNode presumably)

That way we can

1. call Migrate on the whole keyspace
2. force the run

and be assured that all replicas now in the cluster are at the new version, i.e. it's safe to bump the cluster version.
</blockquote></details>

Instead of tracking the range version in the replica state (somewhat out of laziness), I opted to instead introduce a `GCReplicas` RPC to the `Migration` service and wire it in using `EveryNode`. Whenever we're running a ranged Migrate command across the entire keyspace for `vX`, we'll want to GC previous version replicas immediately after. I think almost always this would mean processing the entire the replica GC queue anyway.
As for disallowing previous version snapshots, I think our MLAI safety checks (discussed below) will already prevent it from happening.
___
*[pkg/kv/kvserver/replica_write.go, line 223 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57445#-MNccg5V8aYyKDVIRTho:-MNpAYQp5QwilvKZbJmj:b5jis8f) ([raw file](https://github.com/cockroachdb/cockroach/blob/45ce792773f010ca42422781ab40da54868c1c58/pkg/kv/kvserver/replica_write.go#L223)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Snapshots that don't include the recipient (as specified by replicaID and descriptor in the snap vs the replicaID of the raft instance) are discarded by the recipient, so this should not be an issue. The particular case you mention will have the learner have a higher replicaID than the incoming snapshot, so the snapshot will also be out for that reason. Finally, a snapshot is discarded unless it moves the LAI forward.
</blockquote></details>

Done, incorporated into the comment here.
___
*[pkg/kv/kvserver/batcheval/cmd_migrate.go, line 16 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57445#-MNcVQ5K89TCiMJ7ZJ6l:-MNddU_l7NdVkYJLvWfo:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/45ce792773f010ca42422781ab40da54868c1c58/pkg/kv/kvserver/batcheval/cmd_migrate.go#L16)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

small nit, but I'm generally -1 on unnecessary import renames.
</blockquote></details>

Done.
___
*[pkg/migration/helper.go, line 39 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57445#-MNcWErx40DIgQFQYsmd:-MNpBNnv7Ndw1sCXs4aQ:b-2jhhp9) ([raw file](https://github.com/cockroachdb/cockroach/blob/45ce792773f010ca42422781ab40da54868c1c58/pkg/migration/helper.go#L39)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Should this be exported given that you need to wrap it in a stabilizing retry loop, and thus is a footgun if used from a migration work function?
</blockquote></details>

good idea, Done.
___
*[pkg/migration/helper.go, line 100 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57445#-MNcVx_jASeI1T3qIuEK:-MNpBQjS685uOYp8DJw1:bpb2h7o) ([raw file](https://github.com/cockroachdb/cockroach/blob/45ce792773f010ca42422781ab40da54868c1c58/pkg/migration/helper.go#L100)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

"progress" is a bit of a misnomer, how about "status"
</blockquote></details>

I agree that "status" is better, but for parity with (the now soured upon) `system.jobs`, I reserved the use of "status" for the overall status of the migration ("ongoing/succeeded/failed"). Progress was the closest next thing here that seemed appropriate.
___
*[pkg/migration/helper.go, line 136 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57445#-MNcWQvTDy965IA0ZVB_:-MNdqYCC42LF26L60egj:bkmg4q9) ([raw file](https://github.com/cockroachdb/cockroach/blob/45ce792773f010ca42422781ab40da54868c1c58/pkg/migration/helper.go#L136)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

That's not true. A node can always join just as EveryNode returns. What we have achieved is that there is now a causal happened-before relation between executing something on every node in the cluster and the next node joining the cluster. We managed to run something on all nodes without a new node joining half-way through, which could have allowed it to pick up some state (off one of the existing nodes that hadn't finished processing yet).

For the example of bumping the cluster version, we'll have achieved proving that by the time EveryNode returns, every node in the cluster, current and future, will use the new cluster version.
</blockquote></details>

yea, I had tried elaborating on that in the paragraphs, but we're database people, I should talk in terms of causality. Reworked this entire block.
___
*[pkg/migration/helper.go, line 162 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57445#-MNcY9e59hjUsfZG6S5L:-MNplJCS8JX130ftHeEX:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/45ce792773f010ca42422781ab40da54868c1c58/pkg/migration/helper.go#L162)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

We should do that soon, too. Ideally in the burst of commits coming out of this one!
</blockquote></details>

Done.
___
*[pkg/migration/migrations.go, line 110 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57445#-MNcZyAc6dUxb2W-AYW1:-MNe-mXiBNjFidfRvjW4:b-4ykqp8) ([raw file](https://github.com/cockroachdb/cockroach/blob/45ce792773f010ca42422781ab40da54868c1c58/pkg/migration/migrations.go#L110)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

At 50k ranges, with each range taking ~200ms (a rather conservative assumption but in a wide-area cluster could be true due to hops from the manager to the replicas) to process the Migrate, a block size of 1 is >2.7 hours. Block size 50 is ~3.3 minutes. Block size 200 should easily be in the cards given that each range is doing very little work here, so down to 50s in this case which seems reasonable.
</blockquote></details>

Makes sense, I'll include this back-of-the-envelope calculation in there.
___
*[pkg/migration/migrations.go, line 116 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57445#-MNcdZR6AF0RidVkLMje:-MNeGPL6EHiodNIrVGR4:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/45ce792773f010ca42422781ab40da54868c1c58/pkg/migration/migrations.go#L116)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Yeah, this is a bit of a wart. We do want to reach the first range, but we can't address StartKey. However, `LocalMax` is on r1, so in this case we basically just use  `LocalMax` instead of `StartKey`.

I wonder if we'd get sanity back if we officially let `r1` begin at `LocalMax`, but that is not worth getting into here or probably ever.
</blockquote></details>

Done.
___
*[pkg/migration/migrations.go, line 128 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57445#-MNc_lsa-n2-rAiTDd43:-MNeIzv28lw2Qq6_5l3q:bjqsh8z) ([raw file](https://github.com/cockroachdb/cockroach/blob/45ce792773f010ca42422781ab40da54868c1c58/pkg/migration/migrations.go#L128)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Doesn't `IterateRangeDescriptors` have to handle txn retries? The closure here would need to be idempotent, but the counters aren't. Maybe `IterateRangeDescriptors` should take an `init func()` as well which can reset them.
</blockquote></details>

Ah, good point. I like the `init` idea, done.
___
*[pkg/migration/migrations.go, line 140 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57445#-MNcaCI_E3GtY7kubL0W:-MNpfFALAA-wWHQy-i-k:b3uyj60) ([raw file](https://github.com/cockroachdb/cockroach/blob/45ce792773f010ca42422781ab40da54868c1c58/pkg/migration/migrations.go#L140)):*
> It sounds reasonable to do this as a separate round of EveryNode

Done. I've added a primitive that simply flushes all engines. Seems like a useful property for the migrations infrastructure to provide. 

> we could just extend that to explicitly mandate syncing all engines.

I prefer dead-simple to-the-point APIs, so keeping BumpClusterVersion RPC as is.
___
*[pkg/roachpb/api.go, line 1295 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57445#-MNcb-yv634wrZSR5Yqc:-MNea4kl-MACVlFqufx0:b-femank) ([raw file](https://github.com/cockroachdb/cockroach/blob/45ce792773f010ca42422781ab40da54868c1c58/pkg/roachpb/api.go#L1295)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

I wonder if `MigrateRange` is a better name.
</blockquote></details>

I think I'll keep it as is. We'd technically still be able to send out a ranged Migrate request over arbitrary keyspans, we're only choosing to paginate because it'd be insane not to.
___
*[pkg/roachpb/api.proto, line 1500 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57445#-MNcb7X63Qjck40wzr03:-MNecEFs3PSau7AbSKnT:b-89q77k) ([raw file](https://github.com/cockroachdb/cockroach/blob/45ce792773f010ca42422781ab40da54868c1c58/pkg/roachpb/api.proto#L1500)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Yeah, that's the impression I had as well.
</blockquote></details>

Cool.
___
*[pkg/roachpb/api.proto, line 1502 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57445#-MNcbBaj0hGZ0lVNSZJH:-MNecNVbEPDTNQwjOphz:bkxqt3q) ([raw file](https://github.com/cockroachdb/cockroach/blob/45ce792773f010ca42422781ab40da54868c1c58/pkg/roachpb/api.proto#L1502)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

No, admin requests are really just a way to invoke random code on the leaseholder. Here we want to properly go through the raft pipeline, which admin requests don't.
</blockquote></details>

Makes sense, this TODO was more for aesthetic reasons than anything functional.
___
*[pkg/roachpb/api.proto, line 1509 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57445#-MNcbQ8A8DFqcvncB22-:-MNecZ4J-aTfQiTPLClf:bezi2v6) ([raw file](https://github.com/cockroachdb/cockroach/blob/45ce792773f010ca42422781ab40da54868c1c58/pkg/roachpb/api.proto#L1509)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

incomplete
</blockquote></details>

Looks like I forgot to push out some changes. Done.
___
*[pkg/roachpb/method.go, line 161 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57445#-MNcbWBsEGLeAVKIJ5yI:-MNedul91UL6Mr9Ff3eI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/45ce792773f010ca42422781ab40da54868c1c58/pkg/roachpb/method.go#L161)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

This comment could be better, something like

> Migrate updates the range state to conform to a specified cluster version. It is our main mechanism for phasing out legacy code below Raft.
</blockquote></details>

Done.
___
*[pkg/sql/catalog/systemschema/system.go, line 347 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57445#-MNcbpUc-mvB_HbSAbsX:-MNpBxxT0HqkxkP_5mCL:b-t3w0ub) ([raw file](https://github.com/cockroachdb/cockroach/blob/45ce792773f010ca42422781ab40da54868c1c58/pkg/sql/catalog/systemschema/system.go#L347)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Is this going to be gossiped? (The answer should be "no", I forget what the rules are).
</blockquote></details>

I'm not sure (lazily hoping one of the SQL reviewers can answer).
___
*[pkg/sql/logictest/testdata/logic_test/distsql_agg, line 578 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57445#-MNcc385945SOpGtv5Dw:-MNeizVq0csJxqPH82Pl:b3zawg6) ([raw file](https://github.com/cockroachdb/cockroach/blob/45ce792773f010ca42422781ab40da54868c1c58/pkg/sql/logictest/testdata/logic_test/distsql_agg#L578)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Huh. What are these about?
</blockquote></details>

Who knows, just came up trying to re-write the logictests to include details for the new system table. 
```
make -j1 testlogic TESTFLAGS='--rewrite'
```

I've previously leaned on @yuzefovich for emotional support during these.

**Review by yuzefovich — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57445)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz, @miretskiy, and @tbg)

___
*[pkg/sql/logictest/testdata/logic_test/distsql_agg, line 578 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57445#-MNcc385945SOpGtv5Dw:-MNy0lUG-057m7H9V1fq:b-w3vt85) ([raw file](https://github.com/cockroachdb/cockroach/blob/45ce792773f010ca42422781ab40da54868c1c58/pkg/sql/logictest/testdata/logic_test/distsql_agg#L578)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Who knows, just came up trying to re-write the logictests to include details for the new system table. 
```
make -j1 testlogic TESTFLAGS='--rewrite'
```

I've previously leaned on @yuzefovich for emotional support during these.
</blockquote></details>

I think I fixed it on master, if you rebase, these should go away.

File: pkg/sql/catalog/systemschema/system.go [hunk 1]
```diff
@@ -342,6 +342,18 @@ CREATE TABLE system.sqlliveness (
     expiration       DECIMAL NOT NULL,
   	FAMILY fam0_session_id_expiration (session_id, expiration)
 )`
+
+	MigrationsTableSchema = `
+CREATE TABLE system.migrations (
```
Comment (irfansharif) [inline pkg/sql/catalog/systemschema/system.go:347]: @ajwerner, would love see what you think of this schema.

**Comment by irfansharif** [comment]
Most of this prototype was folded into https://github.com/cockroachdb/cockroach/pull/57650. The contentious opinions here (https://github.com/cockroachdb/cockroach/pull/57662) were scrapped. As for introspection into in-progress migration state, we're simply making use of logging channels for now. There's something to be said for "jobs-ifying" long running migrations at a per-migration level (which in this prototype we simply re-built and dumped into a `system.migrations` table), but that's not a thread we're going to pull on for a bit I think.

**Pull request closed without merging**